aten/src/ATen/native/cuda/Distributions.cu

```diff
@@ -31,28 +41,85 @@ void poisson_cuda_kernel(
     at::Tensor& ret,
     const at::Tensor& lambda,
     std::pair<uint64_t, uint64_t> seeds) {
-  at::cuda::CUDA_tensor_apply2<scalar_t, float>(
+  at::cuda::CUDA_tensor_apply2<scalar_t, scalar_t>(
       ret,
       lambda,
       [seeds] __device__(
-          scalar_t & ret_val, const float& lambda, bool early_exit) {
+          scalar_t & ret_val, const scalar_t& lambda, bool early_exit) {
+        curandStatePhilox4_32_10_t state;
+        curand_init(
+            seeds.first,
+            blockIdx.x * blockDim.x + threadIdx.x,
+            seeds.second,
+            &state);
+        ret_val = scalar_cast<scalar_t>(curand_poisson(&state, scalar_cast<float>(lambda)));
+      });
+}
+
+template <typename scalar_t>
+void gamma_cuda_kernel(
+    at::Tensor& ret,
+    const at::Tensor& alpha,
+    std::pair<uint64_t, uint64_t> seeds) {
+  using accscalar_t = at::cuda::acc_type<scalar_t>;
+  at::cuda::CUDA_tensor_apply2<scalar_t, scalar_t>(
+      ret,
+      alpha,
+      [seeds] __device__(
+          scalar_t & ret_val, const scalar_t& alpha, bool early_exit) {
         curandStatePhilox4_32_10_t state;
         curand_init(
             seeds.first,
             blockIdx.x * blockDim.x + threadIdx.x,
             seeds.second,
             &state);
-        ret_val = scalar_cast<scalar_t>(curand_poisson(&state, lambda));
+        BaseSampler<accscalar_t> standard_uniform([&state] __device__ () {
+          return curand_uniform(&state);
+        });
+        BaseSampler<accscalar_t> standard_normal([&state] __device__ () {
+          return curand_normal(&state);
+        });
+        auto sample = sample_gamma<scalar_t, accscalar_t>(alpha, standard_uniform, standard_normal);
+        ret_val = ((THCNumerics<scalar_t>::min() > sample) ? (THCNumerics<scalar_t>::min()) : sample);
+      });
+}
+
+template <typename scalar_t>
+void gamma_grad_cuda_kernel(
+    at::Tensor& ret,
+    const at::Tensor& self,
+    const at::Tensor& output) {
+  using accscalar_t = at::cuda::acc_type<scalar_t>;
+  at::cuda::CUDA_tensor_apply3<scalar_t, scalar_t, scalar_t>(
+      ret, self, output,
+      [] __device__ (scalar_t& ret_val, const scalar_t& self_val, const scalar_t &output_val) {
+        ret_val = standard_gamma_grad_one<scalar_t, accscalar_t>(self_val, output_val);
       });
 }
+
 } // namespace
 
 namespace at { namespace native {
 Tensor _s_poisson_cuda(const Tensor& lambda, Generator* gen) {
   Tensor ret = lambda.type().tensor(lambda.sizes());
-  auto lambda_ = lambda.toType(ScalarType::Float);
-  AT_DISPATCH_FLOATING_TYPES(ret.type(), "poisson", [&] {
-     poisson_cuda_kernel<scalar_t>(ret, lambda_, next_philox_seed(gen));
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(ret.type(), "poisson", [&] {
+    poisson_cuda_kernel<cuda::type<scalar_t>>(ret, lambda, next_philox_seed(gen, 20));
```

Comment:
@ngimel shouldn't the increment be `#threads * 20` here?

Reply:
The threads will use different sequences, which is separate from offset.